CMakeLists.txt

```diff
@@ -1,5 +1,25 @@
 cmake_minimum_required(VERSION 3.20)  # For using CMAKE_<LANG>_BYTE_ORDER
 
+if(NOT DEFINED USE_VCPKG OR USE_VCPKG)
+  message("NOT DEFINED USE_VCPKG OR USE_VCPKG")
+  if(DEFINED ENV{VCPKG_ROOT})
+    message("DEFINED ENV{VCPKG_ROOT}")
+    if (CMAKE_TOOLCHAIN_FILE)
+      cmake_path(ABSOLUTE_PATH CMAKE_TOOLCHAIN_FILE NORMALIZE OUTPUT_VARIABLE VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
+    endif()
+
+    set(VCPKG_TOOLCHAIN_FILE "scripts/buildsystems/vcpkg.cmake")
+    cmake_path(ABSOLUTE_PATH VCPKG_TOOLCHAIN_FILE BASE_DIRECTORY $ENV{VCPKG_ROOT} NORMALIZE OUTPUT_VARIABLE CMAKE_TOOLCHAIN_FILE)
+
+    if(CMAKE_TOOLCHAIN_FILE STREQUAL VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
+      # prevent endless recursion
```

Comment:
How can recursion happen?

Reply:
`VCPKG_CHAINLOAD_TOOLCHAIN_FILE`, and the added build logic to conditionally set that variable, is ultimately the cause of the recursion problem. 

vcpkg itself requires the use of `CMAKE_TOOLCHAIN_FILE`, which effectively precludes any _other_ use of that variable. In order to "play nice" with other toolchain-based builds, vcpkg provides `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` in order to actually use a different toolchain (ie, for Android or an embedded system).

Gradle (Android's build system) automatically - _and hard-codedly_ - sets `CMAKE_TOOLCHAIN_FILE` on the CMake invocation it kicks off for the C++ compilation side of its overall build. This causes a conflict: `CMAKE_TOOLCHAIN_FILE` will _always_ point to the Android NDK when building for Android, but vcpkg _requires_ that it point to vcpkg.

The solution I landed upon was to have our build file _re_-set the `CMAKE_TOOLCHAIN_FILE` when using vcpkg, copying the passed-in value into the `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` variable vcpkg uses to play nice. All of this happens appropriately considering whether or not we're supposed to be using vcpkg in the first place.

---
**Now, here's where it gets fun:**

`CMakePresets.txt` causes some of these values to be set even when they're not specified on the invocation command line. 

Doing this, with `$ENV{VCPKG_ROOT}` set:
```
cmake --preset mac
```
ends up storing `USE_VCPKG=ON` and `CMAKE_TOOLCHAIN_FILE=/Users/csarbora/Projects/vcpkg/` (for me) in the CMakeCache.txt file, and then those values are already set (as if they were passed in on the command line) for subsequent CMake runs. This is intentional by CMake, so that this works:
```
cmake -B builds/mac -DUSE_VCPKG=OFF -DSOME_OTHER_OPTION=FOO
cmake --build builds/mac 
#            ^^          USE_VCPKG and SOME_OTHER_OPTION will be set even tho they are unspecified
```
When `CMAKE_TOOLCHAIN_FILE` is _passed in_ pointing to vcpkg (because the value is cached from a prior vcpkg-using configuration), `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` gets set to the same value. This would end up causing vcpkg trying to chainload itself, resulting in effectively an infinite recursion. So in order to halt that, I check if the two variables are equal to each other and unset the latter if they are.